evennia/contrib/aws-s3-cdn.py

```diff
@@ -0,0 +1,847 @@
+        self._is_dirty = True
+        if self._multipart is None:
+            self._multipart = self.obj.initiate_multipart_upload(
+                **self._storage._get_write_parameters(self.obj.key)
+            )
+        if self.buffer_size <= self._buffer_file_size:
+            self._flush_write_buffer()
+        bstr = force_bytes(content)
+        self._raw_bytes_written += len(bstr)
+        return super(S3Boto3StorageFile, self).write(bstr)
+
+    @property
+    def _buffer_file_size(self):
+        pos = self.file.tell()
+        self.file.seek(0, os.SEEK_END)
+        length = self.file.tell()
+        self.file.seek(pos)
+        return length
+
+    def _flush_write_buffer(self):
+        """
+        Flushes the write buffer.
+        """
+        if self._buffer_file_size:
+            self._write_counter += 1
+            self.file.seek(0)
+            part = self._multipart.Part(self._write_counter)
+            part.upload(Body=self.file.read())
+            self.file.seek(0)
+            self.file.truncate()
+
+    def _create_empty_on_close(self):
+        """
+        Attempt to create an empty file for this key when this File is closed if no bytes
+        have been written and no object already exists on S3 for this key.
+        This behavior is meant to mimic the behavior of Django's builtin FileSystemStorage,
+        where files are always created after they are opened in write mode:
+            f = storage.open("file.txt", mode="w")
+            f.close()
+        """
+        assert "w" in self._mode
+        assert self._raw_bytes_written == 0
+
+        try:
+            # Check if the object exists on the server; if so, don't do anything
+            self.obj.load()
+        except ClientError as err:
+            if err.response["ResponseMetadata"]["HTTPStatusCode"] == 404:
+                self.obj.put(
+                    Body=b"", **self._storage._get_write_parameters(self.obj.key)
+                )
+            else:
+                raise
+
+    def close(self):
+        if self._is_dirty:
+            self._flush_write_buffer()
+            # TODO: Possibly cache the part ids as they're being uploaded
+            # instead of requesting parts from server. For now, emulating
+            # s3boto's behavior.
+            parts = [{'ETag': part.e_tag, 'PartNumber': part.part_number}
+                     for part in self._multipart.parts.all()]
+            self._multipart.complete(
+                MultipartUpload={'Parts': parts})
+        else:
+            if self._multipart is not None:
+                self._multipart.abort()
+            if 'w' in self._mode and self._raw_bytes_written == 0:
+                self._create_empty_on_close()
+        if self._file is not None:
+            self._file.close()
+            self._file = None
+
+
+@deconstructible
+class S3Boto3Storage(Storage):
+    """
+    Amazon Simple Storage Service using Boto3
```

Comment:
The relationship between this and the `S3Boto3StorageFile` is not clear, nor which of them should be actually used in code (even though it's clear once reading the code that this one is the one actually accessing the settings).